### PR TITLE
refactor!: Rewrote the ambiguity solver for clarity and added Optional Hits Selector

### PR DESCRIPTION
**Pull request by Ragansu**
The Score based solver needed one more feature to be more compatible with Athena, that is optional hit selector. The optional selector enables users to add optional function to the hit selection stage which was previously NOT possible.
In order to do this I had to template the existing getCleanedOutTracks.

This gave me an opportunity to rewrite the whole solver again for efficiency and clarity.

Ragansu Chakkappai.

--- END COMMIT MESSAGE ---

@CarloVarni , @Corentin-Allaire

### CODE REVIEW

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp [hunk 1]
```diff
@@ -80,11 +81,17 @@ class ScoreBasedAmbiguityResolution {
     std::size_t nSharedHits = 0;
   };
 
-  /// @brief MeasurementInfo : contains the measurement ID and the detector ID
-  struct MeasurementInfo {
-    std::size_t iMeasurement = 0;
-    std::size_t detectorId = 0;
-    bool isOutlier = false;
+  enum class TrackStateTypes : int {
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:84]: I suppose se can also go for `std::uint8_t`?

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp [hunk 2]
```diff
@@ -80,11 +81,17 @@ class ScoreBasedAmbiguityResolution {
...
+    // A measurement not yet used in any other track
+    UnsharedHit = 1,
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:86]: why you start with 1?

Comment (CarloVarni) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:86]: do you really need to specify the number?

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp [hunk 3]
```diff
@@ -21,27 +21,16 @@ inline const Logger& ScoreBasedAmbiguityResolution::logger() const {
...
-  auto MeasurementIndexMap =
-      std::unordered_map<SourceLink, std::size_t, source_link_hash_t,
-                         source_link_equality_t>(0, sourceLinkHash,
-                                                 sourceLinkEquality);
-
-  std::vector<std::vector<MeasurementInfo>> measurementsPerTrack;
-  measurementsPerTrack.reserve(tracks.size());
+    const track_container_t& tracks) const {
   ACTS_VERBOSE("Starting to compute initial state");
+  std::vector<std::vector<TrackFeatures>> trackFeaturesVectors = {};
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:29]: maybe you can call a reserve on this? like the number of tracks?

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp [hunk 4]
```diff
@@ -440,25 +402,194 @@ std::vector<int> Acts::ScoreBasedAmbiguityResolution::solveAmbiguity(
...
+  // Stores tracks measurement into a vector or vectors
+  // (measurementsPerTrackVector) and counts the number of tracks per
+  // measurement.(nTracksPerMeasurement)
+
+  for (const auto& track : tracks) {
+    std::vector<std::size_t> measurementsPerTrack;
+    for (const auto& ts : track.trackStatesReversed()) {
+      if (ts.typeFlags().test(Acts::TrackStateFlag::OutlierFlag) ||
+          ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        ACTS_DEBUG("Track state type is OutlierFlag");
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:422]: this is not true. You also allow for measurements.

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp [hunk 5]
```diff
@@ -440,25 +402,194 @@ std::vector<int> Acts::ScoreBasedAmbiguityResolution::solveAmbiguity(
...
+  std::map<std::size_t, std::size_t> nTracksPerMeasurement;
+
+  // Stores tracks measurement into a vector or vectors
+  // (measurementsPerTrackVector) and counts the number of tracks per
+  // measurement.(nTracksPerMeasurement)
+
+  for (const auto& track : tracks) {
+    std::vector<std::size_t> measurementsPerTrack;
+    for (const auto& ts : track.trackStatesReversed()) {
+      if (ts.typeFlags().test(Acts::TrackStateFlag::OutlierFlag) ||
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:420]: I'd suggest the logic 
```cpp
if (not ...) continue;
...
```
in order to reduce the number of nested `{}`

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp [hunk 6]
```diff
@@ -440,25 +402,194 @@ std::vector<int> Acts::ScoreBasedAmbiguityResolution::solveAmbiguity(
...
+          ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        ACTS_DEBUG("Track state type is OutlierFlag");
+        Acts::SourceLink sourceLink = ts.getUncalibratedSourceLink();
+        // assign a new measurement index if the source link was not seen yet
+        auto emplace = MeasurementIndexMap.try_emplace(
+            sourceLink, MeasurementIndexMap.size());
+        auto iMeasurement = emplace.first->second;
```
Comment (CarloVarni) [inline Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:427]: can you be explicit with the types instead of using `auto`?

**Comment by CarloVarni** [comment]
@Ragansu do I understand that right that it is also computing and labeling hits as `SharedHits`?

**Review by CarloVarni — CHANGES_REQUESTED** [review]

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp (hunk 1 above)
Comment (Ragansu) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:84]: yes uint_8  should work

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp (hunk 2 above)
Comment (Ragansu) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:86]: Starting with one is a convention taken from Athena,  I am not sure we need a specific numbers, But I haven't tested without it.

Comment (CarloVarni) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.hpp:86]: I believe we can simply introduce a conversion function in Athena that converts this enum to the corresponding Athena enum. We do the same for the OutputLevel. 

I'd simplify this by not specifying the actual value. The `class` will forbid implicit conversion to/from integers alsready

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp (hunk 4 above)
Comment (Ragansu) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:422]: Oh yes. that's a mistake, I will fix it.

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp [hunk 7]
```diff
@@ -440,25 +402,194 @@ std::vector<int> Acts::ScoreBasedAmbiguityResolution::solveAmbiguity(
...
+    const std::vector<
+        std::function<void(const track_proxy_t&,
+                           const typename track_proxy_t::ConstTrackStateProxy&,
+                           TrackStateTypes&)>>& optionalHitSelections) const {
+  // For tracks with shared hits, we need to check and remove bad hits
+
+  std::vector<TrackStateTypes> trackStateTypes;
+  // Loop over all measurements of the track and for each hit a
+  // trackStateTypes is assigned.
+  for (std::size_t index = 0; auto ts : track.trackStatesReversed()) {
```
[the remark this answers is not in the document]
Comment (Ragansu) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp:498]: const auto& should work, I think I will test it and let you know.

**Comment by Ragansu** [comment]
> @Ragansu do I understand that right that it is also computing and labeling hits as `SharedHits`?

Yes and No, There is computing of number of tracks shared with a said hit, but its not assigning to `Acts::TrackStateFlag`
We still need to add the `Acts::TrackStateFlag::SharedHitFlag` before the track reaches ambi solver, 

I could rewrite the code to include this calculation into the ambi solver but I believe it should be part of the original const track object. So before ambiguity solver.

File: Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp (hunk 6 above)
Comment (CarloVarni) [reply Core/include/Acts/AmbiguityResolution/ScoreBasedAmbiguityResolution.ipp]: it's fine then. don't worry

**Comment by CarloVarni** [comment]
@Ragansu can you fix the docs test?

**Comment by Ragansu** [comment]
> @Ragansu can you fix the docs test? (assuming it is still there ...)

You mean the one shown by sonar cube ? I probably can but it will take decent amount of time to make appropriate unit tests. 
Can I do it in a later PR?

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [a13c4adaf8935e1ba8d13bb83f2d9ad90f596459]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8545842)!

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193129)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193128) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_run4_acts_vertex_PU200/pulls/3805/)|
|:green_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193127) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_run4_ttbar_PU200/pulls/3805/)|
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193126) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_data18_13TeV_1000evt/pulls/3805/)|
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193125) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_ttbarPU40_reco/pulls/3805/)|
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193124)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193123)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193122)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193121)||
|:red_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193120)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193119)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflowHgtd](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193118)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193117)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193116)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193115)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193114)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193113)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193112)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193111)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193110)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193109)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193108)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193107)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193106)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193105)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193104)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193103)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193102)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193101)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193100)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193099)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193098)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193097)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193096)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193095)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46193094)||

**Comment by CarloVarni** [comment]
ok we good to go on this one

**Review by CarloVarni — APPROVED** [review]

**Pull request merged**